Packet++/src/Sll2Layer.cpp

```diff
@@ -0,0 +1,212 @@
+
+	return true;
+}
+
+void Sll2Layer::parseNextLayer()
+{
+	if (m_DataLen <= sizeof(sll2_header))
+		return;
+
+	uint8_t* payload = m_Data + sizeof(sll2_header);
+	size_t payloadLen = m_DataLen - sizeof(sll2_header);
+
+	sll2_header* hdr = getSll2Header();
+	switch (be16toh(hdr->protocol_type))
+	{
+	case PCPP_ETHERTYPE_IP:
+		m_NextLayer = IPv4Layer::isDataValid(payload, payloadLen)
+			? static_cast<Layer*>(new IPv4Layer(payload, payloadLen, this, m_Packet))
+			: static_cast<Layer*>(new PayloadLayer(payload, payloadLen, this, m_Packet));
+		break;
+	case PCPP_ETHERTYPE_IPV6:
+		m_NextLayer = IPv6Layer::isDataValid(payload, payloadLen)
+			? static_cast<Layer*>(new IPv6Layer(payload, payloadLen, this, m_Packet))
+			: static_cast<Layer*>(new PayloadLayer(payload, payloadLen, this, m_Packet));
+		break;
+	case PCPP_ETHERTYPE_ARP:
+		m_NextLayer = new ArpLayer(payload, payloadLen, this, m_Packet);
+		break;
+	case PCPP_ETHERTYPE_VLAN:
+	case PCPP_ETHERTYPE_IEEE_802_1AD:
+		m_NextLayer = new VlanLayer(payload, payloadLen, this, m_Packet);
+		break;
+	case PCPP_ETHERTYPE_PPPOES:
+		m_NextLayer = PPPoESessionLayer::isDataValid(payload, payloadLen)
+			? static_cast<Layer*>(new PPPoESessionLayer(payload, payloadLen, this, m_Packet))
+			: static_cast<Layer*>(new PayloadLayer(payload, payloadLen, this, m_Packet));
+		break;
+	case PCPP_ETHERTYPE_PPPOED:
+		m_NextLayer = PPPoEDiscoveryLayer::isDataValid(payload, payloadLen)
+			? static_cast<Layer*>(new PPPoEDiscoveryLayer(payload, payloadLen, this, m_Packet))
+			: static_cast<Layer*>(new PayloadLayer(payload, payloadLen, this, m_Packet));
+		break;
+	case PCPP_ETHERTYPE_MPLS:
+		m_NextLayer = new MplsLayer(payload, payloadLen, this, m_Packet);
+		break;
+	default:
+		m_NextLayer = new PayloadLayer(payload, payloadLen, this, m_Packet);
+	}
+
+}
+
+void Sll2Layer::computeCalculateFields()
+{
+	if (m_NextLayer == nullptr)
+		return;
+
+	sll2_header* hdr = getSll2Header();
+	switch (m_NextLayer->getProtocol())
+	{
+		case IPv4:
+			hdr->protocol_type = htobe16(PCPP_ETHERTYPE_IP);
+			break;
+		case IPv6:
+			hdr->protocol_type = htobe16(PCPP_ETHERTYPE_IPV6);
+			break;
+		case ARP:
+			hdr->protocol_type = htobe16(PCPP_ETHERTYPE_ARP);
+			break;
+		case VLAN:
+			hdr->protocol_type = htobe16(PCPP_ETHERTYPE_VLAN);
+			break;
+		default:
+			return;
+	}
+}
+
+bool Sll2Layer::isDataValid(const uint8_t* data, size_t dataLen)
+{
+	return data && dataLen >= sizeof(sll2_header);
+}
+
+std::string Sll2Layer::toString() const
+{
+	return "Linux cooked header v2";
+}
+
+uint16_t Sll2Layer::getProtocolType() const
+{
+	return be16toh(getSll2Header()->protocol_type);
+}
+
+void Sll2Layer::setProtocolType(uint16_t protocolType)
+{
+	getSll2Header()->protocol_type = htobe16(protocolType);
+}
+
+uint16_t Sll2Layer::getReservedType() const
+{
+	return be16toh(getSll2Header()->reserved_type);
+}
+
+void Sll2Layer::setReservedType(uint16_t reservedType)
+{
+	getSll2Header()->reserved_type = htobe16(reservedType);
+}
```

Comment:
These 3 methods (`setProtocolType()`, `getReservedType()`, `setReservedType()`) also don't have tests, can you add tests for them?